### ISSUE TITLE
RFE: Combine Damaged Lances into a Single Lance

### ISSUE THREAD

**Issue by Degrath** [MegaMek/mekhq#484]
the ability in the TO&E list to have a force block for each mek in a lance for assigning a tech to each mech in a force without assigning to a specific mek. currently 1 tech can be assigned per force. Also if a company temporarily drops in strength while on mission and you lack the fund to replace personnel the campaign could recognize 2 separate lances as campaign eligible by the next higher parent force without having to manually shift people around.

I think I am nearly 1/2 way there in the Lance.java code but I cant seem to get it to read a force block with no units inside even though there are subforce units in other force blocks. If I place a unit into the force block it will register that unit plus all subforce units but not without.
 Here is what I changed to get it to work somewhat:

`public int getSize(Campaign c) {`
`        if (c.getFaction().isClan()) {`
`           return (int)Math.ceil(getEffectivePoints(c));`
`        }`
`        if (c.getForce(forceId) != null) {`
`        	//THIS is WHAT I CHANGED .getUnits to .getAllUnits`
`            return c.getForce(forceId).getAllUnits().size();`
`        } else {`
`            return 0;`
`        }`
I also changed it in a few other place like `public double getEffectivePoints(Campaign c){` function where .getUnits() was used to .getAllUnits()

Where I am stuck is:

`public boolean isEligible(Campaign c) {`
`        /* Check that the number of units and weight are within the limits`
`         * and that the force contains at least one ground unit. */`
`       if (c.getCampaignOptions().getLimitLanceNumUnits()) {`
`            int size = getSize(c);`
`            if (size < getStdLanceSize(c.getFaction()) - 1 ||`
`                    size > getStdLanceSize(c.getFaction()) + 2) {`
`                return false;`
`            }`
`        }`
`        if (c.getCampaignOptions().getLimitLanceWeight() &&`
`                getWeightClass(c) > EntityWeightClass.WEIGHT_ASSAULT) {`
`            return false;`
`        }`
`        boolean hasGround = false;`
`      //THIS is WHAT I CHANGED .getUnits to .getAllUnits`
`        for (UUID id : c.getForce(forceId).getAllUnits()) {`
`            Unit unit = c.getUnit(id);`
`            if (null != unit) {`
`                Entity entity = unit.getEntity();`
`                if (null != entity) {`
`                    if (UnitType.determineUnitTypeCode(entity) >= UnitType.JUMPSHIP) {`
`                        return false;`
`                    }`
`                    if ((entity.getEntityType() & ETYPE_GROUND) != 0) {`
`                        hasGround = true;`
`                    }`
`                }`
`            }`
`        }`
`        return hasGround;`
`    }`

The first part of the check makes sure a lance conforms with the faction required numbers +/- a few
IS 3-5, Clan 4-7, etc... Removing the check I get each force that meets the criteria for the second and third portions of this function for weight and hasGround as each individual lance subforce unit but not as the parent lance force. I was thinking a for loop through each of the force blocks but so far have not been able to make it happen.

I understand this level of micromanaging would not be for everyone so I have been trying to get it on my own with only basic java skills. An option to disable it would in campaign options would be needed if I could get it as a feature request. or possibly a nudge in the right direction on getting this loop to work.
So far I have noticed no ill effects while running MekHq with the changes I have made to the code.

**Comment by HammerGS** [comment]
One of the devs that code will see this. But if you join our Slack channel their are often devs and contributors to the project that can help and give advice. 

http://megamek.org:3000/

**Comment by Degrath** [comment]
thanks for the info and prompt reply

**Comment by Bonepart** [comment]
It seems like some amount of paperwork is necessary and should be required. If I'm reading this right, you want to be able to set up say a Company, with two Lances, each with four mechs. Say two mechs in each Lance is destroyed, you want it to treat your two half strength Lances as a full Lance and let you deploy it.

That seems very wrong, at least to me.

For the other I'm not entirely sure what you are asking for. Maybe the ability to set a pool of techs? Say assign four techs to this Lance and they will get assigned to a mech automatically as mechs are moved in and out of the Lance?

**Comment by Degrath** [comment]
> Say two mechs in each Lance is destroyed, you want it to treat your two half strength Lances as a full Lance and let you deploy it.
 On the battlefield this kind of thing happens all the time, an unofficial cobbling of forces until reinforcements can be brought in to bring the units back to full strength without compromising future unit integrity.
The other reason is the ability to assign a tech to a specific unit slot in a lance and have it in the TO&E to change as a mek may be changed in a lance slot, currently possible from the personnel window to assign a tech to a mek yes but when you start to get lots of units and tech persons the personnel window gets harder to make sense of even with filtering for some, maybe not all.
Lastly the way I personally break down a unit and name companies/lances/units it fits my need for micromanagement. It is why I started with my own solution by altering the code. I know this isn't for everyone and might not make sense the way I am able to explain it.  It does not have to become a feature, I just need a nudge in the right direction to alter why it ignores an empty force box with with any unit of any type in the box it will read all subforce boxes. Possibly as I get it figured out I can tie it to an option in the campaign options but right now with my basic abilities in java I am not there yet.

**Comment by Degrath** [comment]
So I am unable to find a way to stop mekhq from ignoring my empty force block and adding all the units from nested subforce blocks below it. Anyone have any ideas?
I make 1 parent force with 4 subforces with 1 unit per subforce, they will not show as campaign eligible
If I add 1 unit of any type to the parent the entire force of units now become eligible. I have looked through Lance.java and Force.java and this think it has something to do with the 
`public boolean isEligible(Campaign c)` function in the first part where is checks a force against the faction and campaign limits. I have been trying to get it to loop through each of force ids but it still does not see empty force blocks.

**Comment by Qwertronix** [comment]
Soooo to get this straight, suppose I have the following TO&E:

- First Company
    - Alpha Lance
        - Mech 1
        - Mech 2
        - Mech 3
        - Mech 4
    - Beta Lance
        - ~~Mech 5~~
        - ~~Mech 6~~
        - Mech 7
        - Mech 8
    - Gamma Lance
        - Mech 9
        - Mech 10
        - ~~Mech 11~~
        - ~~Mech 12~~

You wouldn't be able to deploy "First Company" as a whole nor Beta nor Gamma, because each has either too few or too many total units, but you'd like to be able to shift things around to simply this: 

- First Company
    - Alpha Lance
        - Mech 1
        - Mech 2
        - Mech 3
        - Mech 4
    - Temporary Lance
        - Beta Lance
            - ~~Mech 5~~
            - ~~Mech 6~~
            - Mech 7
            - Mech 8
        - Gamma Lance
            - Mech 9
            - Mech 10
            - ~~Mech 11~~
            - ~~Mech 12~~

and be able to assign "Temporary Lance" as a Fight/Scout/etc. unit even though the root force "Temporary Lance" itself has no unit directly inside it, but checking the sub-forces (Beta and Gamma in this case) to see if they match the requirements, in order to keep Beta and Gamma organized?

**Comment by Degrath** [comment]
@Qwertronix  yes thats is part of it. I do however see an issue you have presented. Without creating a temp lance subforce the next higher force would still be invalid because lance 1 would be operational with 4 units and 2 from lance 2 and 3 would make the parent force = 8 units which is outside maximum units defined in the code. The current code for making a force valid for a contract is 2 units of any type with at least 1 "ground" unit for a total of 3 minimum.

     Part of the benefit to this for me is when dealing with a large unit I can change a unit from the TOE without having to reassign the tech also if 1 or more unit is disabled in each lance the next higher force unit could still be valid without manually reorganizing and then having to change back when replacement units come online. 

     Expanding the functionality of the TOE "Force" icon with the ability to attach a medical officer and administration staff could be useful in compartmentalizing a unit to handle multiple contracts on different planets. Also allowing for units like Aerospace and infantry to be valid for deploying a single flight(IS = 2 aeros) for aerospace missions or attaching a platoon of infantry with a vehicle / mech lance. A single dropship or jumpship should be a valid force for people who like to run space engagements prior to a planetary landing.

**Comment by Qwertronix** [comment]
> @Qwertronix yes thats is part of it. I do however see an issue you have presented. Without creating a temp lance subforce the next higher force would still be invalid because lance 1 would be operational with 4 units and 2 from lance 2 and 3 would make the parent force = 8 units which is outside maximum units defined in the code. The current code for making a force valid for a contract is 2 units of any type with at least 1 "ground" unit for a total of 3 minimum.
Part of the benefit to this for me is when dealing with a large unit I can change a unit from the TOE without having to reassign the tech also if 1 or more unit is disabled in each lance the next higher force unit could still be valid without manually reorganizing and then having to change back when replacement units come online.

Yeah, this can be changed to work like I suggested, where the temporary force would still be a manual adjustment on the player's part but should be a lot less tedious housekeeping than reassigning units to other forces and readjusting all the assigned technicians. 

> Expanding the functionality of the TOE "Force" icon with the ability to attach a medical officer and administration staff could be useful in compartmentalizing a unit to handle multiple contracts on different planets. Also allowing for units like Aerospace and infantry to be valid for deploying a single flight(IS = 2 aeros) for aerospace missions or attaching a platoon of infantry with a vehicle / mech lance. A single dropship or jumpship should be a valid force for people who like to run space engagements prior to a planetary landing.

This is a whoole other beast I don't even want to approach and would need to be part of a far more massive feature upgrade to several subsystems.

**Comment by IllianiCBT** [comment]
This would require significant rework of multiple deep, core systems. I'm not convinced the benefit warrants the effort and likelihood of cascading problems.

Given the age of this request and that there hasn't been any movement in several years I'm going to go ahead and close it.
